engine/helper/pka_helper.c

```diff
@@ -721,16 +723,132 @@ static void set_bignum(pka_bignum_t *bn, pka_operand_t *operand)
 }
 
 //
-// Synchronous PKA implementation
+// Synchronous/Asynchronous PKA implementation
 //
 
+
+struct poll_args {
+    ASYNC_JOB     *job;
+    int            fd;
+    pka_operand_t *operand;
+    ecc_point_t  **ecc_point;
+};
+
+
+void init_operand(pka_operand_t *operand,
+                  uint8_t       *buf,
+                  uint32_t       buf_len,
+                  uint8_t        big_endian)
+{
+    memset(operand, 0, sizeof(pka_operand_t));
+    memset(buf,     0, buf_len);
+    operand->buf_ptr    = buf;
+    operand->buf_len    = buf_len;
+    operand->actual_len = 0;
+    operand->big_endian = big_endian;
+}
+
+
+typedef enum { PKA_SYNC, PKA_ASYNC, PKA_ASYNC_ERROR_PIPE,
+               PKA_ASYNC_ERROR_OPERAND, PKA_ASYNC_ERROR_ARGS } pka_mode_t;
+
+typedef enum { PKA_ALGO_MOD, PKA_ALGO_ECC } pka_algo_t;
+
+
+pka_mode_t init_pka_async_job(pka_handle_t handle,     pka_algo_t    algo,
+                              pka_operand_t **operand, ecc_point_t **ecc_point)
+{
+    ASYNC_JOB *job = ASYNC_get_current_job();
+
+    if (job)
+    {
+        int               pipefds[2] = {0, 0}, fd = 0;
+        struct poll_args *p_args = NULL;
```

Comment:
Regarding mixing up variable definition/declaration and code, it is common practice when it comes to kernel development. If there is a need for an extra context or a separate stack then it is better to break the function. Again I am just pointing this out because from readability point of view the variable scope is not obvious. You may disagree though :)

Anyway I will leave it you. Use your best judgement ;)

(by the way I can no longer see my previous comment)